rust/core-lib/src/width_cache.rs

```diff
@@ -19,44 +19,93 @@ use std::collections::{BTreeMap, HashMap};
 
 use xi_rpc;
 
-use client::{Client, WidthReq};
+use client::Client;
 
 /// A token which can be used to retrieve an actual width value when the
 /// batch request is submitted.
 ///
 /// Internally, it is implemented as an index into the `widths` array.
 pub type Token = usize;
 
+/// A measured width, in points.
+type Width = f64;
+
+type StyleId = usize;
+
 pub struct WidthCache {
-    // maps cache key to index within widths
+    /// maps cache key to index within widths
     m: HashMap<WidthCacheKey<'static>, Token>,
-    widths: Vec<f64>,
+    widths: Vec<Width>,
 }
 
 #[derive(Eq, PartialEq, Hash)]
 struct WidthCacheKey<'a> {
-    id: usize, // style id
+    id: StyleId,
     s: Cow<'a, str>,
 }
 
 /// A batched request, so that a number of strings can be measured in a
 /// a single RPC.
 pub struct WidthBatchReq<'a> {
     cache: &'a mut WidthCache,
-    pending_tok: usize,
+    pending_tok: Token,
     req: Vec<WidthReq>,
     req_toks: Vec<Vec<Token>>,
     // maps style id to index into req/req_toks
-    req_ids: BTreeMap<usize, usize>,
+    req_ids: BTreeMap<StyleId, Token>,
+}
+
+/// A request for measuring the widths of strings all of the same style
+/// (a request from core to front-end).
+#[derive(Serialize, Deserialize)]
+pub struct WidthReq {
+    pub id: StyleId,
+    pub strings: Vec<String>,
+}
+
+/// The response for a batch of [`WidthReq`]s.
+pub type WidthResponse = Vec<Vec<Width>>;
+
+/// A trait for types that provide width measurement. In the general case this
+/// will be provided by the frontend, but alternative implementations might
+/// be provided for faster measurement of 'fixed-width' fonts, or for testing.
+pub trait Measure {
```

Comment:
Odd for me to prefer more verbose names, but WidthMeasure feels better here. I don't feel strongly, though.